server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconciler.java

```diff
@@ -526,7 +545,7 @@ private void balance() {
                     continue;
                 }
 
-                undesiredAllocations++;
+                undesiredAllocationsPerNode.compute(shardRouting.currentNodeId(), (id, count) -> count == null ? 1 : count + 1);
```

Comment:
I wonder if instead we could increment only if the `shardRouting.currentNodeId()` is not in `allocation.metadata().nodeShutdowns()`?

This will be the same hash-lookup and will not require additional data structure.